### PR TITLE
FEAT: Add port close command feature and improve error handling

### PR DESCRIPTION
**Pull request by arcker**
This commit implements the port close command feature that allows users to close previously opened ports. Changes include:

- Add close command implementation in client side
- Implement close_port method in firewall implementation
- Update protocol handling to support Close operation
- Add robust error handling for edge cases
- Implement detailed debug logging for better diagnostics
- Update test scripts to verify close command functionality with both IPv4 and IPv6
- Fix assertions to accept both Knock and Close operations

The implementation follows the same security patterns as the knock command with proper authentication.

### CODE REVIEW

File: letmein-proto/src/lib.rs [hunk 1]
```diff
@@ -676,7 +684,7 @@ mod tests {
     fn test_msg_raw_invalid_operation() {
```
Comment (mbuesch) [inline letmein-proto/src/lib.rs:684]: Please also add a new test case for `Close` similar to `test_msg_knock`.

File: letmein/src/command/close.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
+// -*- coding: utf-8 -*-
```
Comment (mbuesch) [inline letmein/src/command/close.rs:1]: I am wondering, if we can unify the close and knock implementations on the client side and create one common "Protocol sequence" implementation for both.
They are basically identical. The only difference should only be the `Operation::Close`.

File: letmein/src/main.rs [hunk 3]
```diff
@@ -70,6 +71,89 @@ fn parse_user(s: &str) -> ah::Result<UserId> {
 
 #[derive(Subcommand, Debug)]
 enum Command {
+    /// Close a previously opened port on a server.
+    Close {
```
Comment (mbuesch) [inline letmein/src/main.rs:75]: Can we avoid the duplication of the struct content?

File: letmeind/src/protocol.rs [hunk 4]
```diff
@@ -121,8 +121,28 @@ impl<'a, C: ConnectionOps> Protocol<'a, C> {
         self.resource_id = None;
         self.auth_state = AuthState::NotAuth;
 
-        // Receive the initial knock message.
-        let knock = self.recv_msg(Operation::Knock).await?;
+        // Receive the initial message (Knock or Close).
+        let initial_msg = match timeout(self.conf.control_timeout(), self.conn.recv_msg())
```
Comment (mbuesch) [inline letmeind/src/protocol.rs:125]: Please change `self.recv_msg` instead of open-coding the timeout and expected operation check here.
We could use a reference to a slice of expected operations as a parameter to `self.recv_msg`.

File: letmeind/src/protocol.rs [hunk 5]
```diff
@@ -121,8 +121,28 @@ impl<'a, C: ConnectionOps> Protocol<'a, C> {
...
+                "Invalid initial message operation. Expected Knock or Close, got {:?}",
+                initial_msg.operation()
+            ));
+        }
+        
+        // Store the operation type before moving initial_msg
+        let operation = initial_msg.operation();
+        let knock = initial_msg;
 
         let user_id = knock.user();
```
Comment (mbuesch) [inline letmeind/src/protocol.rs:147]: This looks misnamed now. It's not always a knock.

File: letmeind/src/protocol.rs [hunk 6]
```diff
@@ -217,14 +237,26 @@ impl<'a, C: ConnectionOps> Protocol<'a, C> {
...
+                if operation == Operation::Close {
+                    // Close port operation
+                    if let Err(e) = fw
+                        .close_port(self.conn.peer_addr().ip(), port_type, *port)
+                        .await
+                    {
+                        let _ = self.send_go_away().await;
+                        return Err(err!("letmeinfwd firewall close: {e}"));
+                    }
+                } else {
```
Comment (mbuesch) [inline letmeind/src/protocol.rs:251]: Please explicitly check for `else if operation == Operation::Knock` here for defensive coding.
The else branch could be made `unreachable!()`.

File: letmeinfwd/src/firewall/nftables.rs [hunk 7]
```diff
@@ -188,14 +188,45 @@ struct ListedRuleset<'a> {
 impl ListedRuleset<'_> {
     /// Get the active ruleset from the kernel.
     pub async fn from_kernel(conf: &Config) -> ah::Result<Self> {
-        let ruleset = get_current_ruleset_with_args_async(
+        println!("firewall: Retrieving current ruleset from kernel...");
```
Comment (mbuesch) [inline letmeinfwd/src/firewall/nftables.rs:191]: Please remove these debug messages and all debug code

File: letmeinfwd/src/firewall/nftables.rs [hunk 8]
```diff
@@ -209,27 +240,38 @@ impl ListedRuleset<'_> {
         port: SingleLeasePort,
     ) -> ah::Result<u32> {
         let comment = gen_rule_comment(Some(addr), port)?;
+        println!("firewall: Looking for rule with comment: '{}'", comment);
+        println!("firewall: Rules found in kernel:");
+        let mut found_any = false;
+        
         for obj in &*self.objs {
```
Comment (mbuesch) [inline letmeinfwd/src/firewall/nftables.rs:247]: These changes look like debug-only, right? If so, please remove these changes.

File: letmeinfwd/src/firewall/nftables.rs [hunk 9]
```diff
@@ -407,7 +482,19 @@ impl NftFirewall {
     async fn nftables_remove_leases(&mut self, conf: &Config, leases: &[Lease]) -> ah::Result<()> {
         if !leases.is_empty() {
             // Get the active ruleset from the kernel.
-            let ruleset = ListedRuleset::from_kernel(conf).await?;
+            let ruleset = match ListedRuleset::from_kernel(conf).await {
+                Ok(ruleset) => ruleset,
+                Err(e) => {
+                    // In debug mode, handle the error gracefully
+                    if conf.debug() {
```
Comment (mbuesch) [inline letmeinfwd/src/firewall/nftables.rs:489]: This breaks backward compatibility. Debug mode is not supposed to change behavior.
If there is a good reason to ignore some errors, these cases should be explicitly coded and documented. It shall be preferred to not be dependent on the configuration whether errors are ignored or not at this level.

File: letmeinfwd/src/firewall/nftables.rs [hunk 10]
```diff
@@ -474,6 +569,51 @@ impl FirewallOpen for NftFirewall {
...
+        println!("firewall: Current leases in memory:");
+        for (addr, p) in self.leases.keys() {
+            println!("firewall:   Lease for {addr} port {p}");
+        }
+        
+        let id = (remote_addr, port);
+        if let Some(lease) = self.leases.remove(&id) {
+            println!("firewall: Found lease in memory for {remote_addr} port {port}, attempting to remove from kernel");
+            // Lease exists, remove it from the kernel
+            let leases = vec![lease];
```
Comment (mbuesch) [inline letmeinfwd/src/firewall/nftables.rs:595]: This doesn't need to be a heap allocating vec.

File: letmeinfwd/src/firewall/nftables.rs [hunk 11]
```diff
@@ -474,6 +569,51 @@ impl FirewallOpen for NftFirewall {
...
+            
+            // Remove lease from kernel (error handling is done in nftables_remove_leases)
+            self.nftables_remove_leases(conf, &leases).await?;
+            self.print_total_rule_count(conf);
+            println!("firewall: Successfully removed lease for {remote_addr} port {port}");
+        } else {
+            println!("firewall: No lease found in memory for {remote_addr} port {port}");
+            
+            // Try to remove from kernel anyway
```
Comment (mbuesch) [inline letmeinfwd/src/firewall/nftables.rs:604]: Nope. Why do this?

File: letmeinfwd/src/server.rs [hunk 12]
```diff
@@ -136,6 +136,60 @@ impl FirewallConnection {
                     self.send_msg(&FirewallMessage::new_nack()).await?;
                 }
             }
+            FirewallOperation::Close => {
```
Comment (mbuesch) [inline letmeinfwd/src/server.rs:139]: This looks like Open and Close could share a lot of code.

**Review by mbuesch — CHANGES_REQUESTED** [review]
In general this looks to be the right approach.
But the code seems to include quite some new hacks/debug/duplications.
